src/Common/GuardedPoolAllocatorOptions.cpp

```diff
@@ -0,0 +1,273 @@
+#include <Common/GuardedPoolAllocatorOptions.h>
+#include <Common/IO.h>
+
+#include <unistd.h>
+
+#include <cassert>
+#include <cstring>
+#include <cstdint>
+#include <cinttypes>
+#include <cstdlib>
+
+
+#include <cstdio>
+
+namespace
+{
+
+enum class OptionType : uint8_t
+{
+    OT_bool,
+    OT_int,
+};
+
+class OptionParser
+{
+public:
+    explicit OptionParser() { }
+    void registerOption(const char * name, const char * desc, OptionType type, void * var);
+    bool parseString(const char * string);
+    void printOptionDescriptions();
+
+private:
+    /// Calculate at compile-time how many options are available.
+#define GWP_ASAN_OPTION(...) +1
+    static constexpr size_t max_options = 0
+#include <Common/GuardedPoolAllocatorOptions.inc>
+        ;
+#undef GWP_ASAN_OPTION
+
+    struct Option
+    {
+        const char * name;
+        const char * desc;
+        OptionType type;
+        void * var;
+    } options[max_options];
+
+    size_t options_total = 0;
+    const char * buffer = nullptr;
+    uintptr_t pos = 0;
+
+    void skipWhitespace();
+    bool parseOptions();
+    bool parseOption();
+    bool setOptionToValue(const char * name, const char * value);
+};
+
+#define ARRAY_SIZE(a) (sizeof(a)/sizeof((a)[0]))
+
+/// Macros to avoid using strlen(), since it may fail if SSE is not supported.
+#define writeError(data) do \
```

Comment:
Initial implementation indeed used printfs instead of this, but @alexey-milovidov discouraged using printf here as well as any other memory allocation dependant options, advising instead to either remove this help message altogether, or to make it static using only write. I decided to still leave this help message since I think it may be useful, so I resorted to this (not so beautiful) use of maco, which I borrowed from programs/main.cpp.